unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs

```diff
@@ -0,0 +1,158 @@
+            IInternalECSComponent<InternalUiContainer> internalUiContainerComponent,
+            IList<IParcelScene> loadedScenes,
+            IWorldState worldState)
+        {
+            var state = new State()
+            {
+                uiDocument = uiDocument,
+                internalUiContainerComponent = internalUiContainerComponent,
+                worldState = worldState,
+                loadedScenes = loadedScenes,
+                lastSceneId = null,
+                isPendingSceneUI = false,
+                currentScene = null
+            };
+
+            return () => Update(state);
+        }
+
+        private static void Update(State state)
+        {
+            string currentSceneId = state.worldState.GetCurrentSceneId();
+            bool sceneChanged = state.lastSceneId != currentSceneId;
+            state.lastSceneId = currentSceneId;
+
+            ApplyParenting(state.internalUiContainerComponent);
+
+            // clear UI if scene changed
+            if (sceneChanged)
+            {
+                ClearUI(state.uiDocument);
+                state.isPendingSceneUI = !string.IsNullOrEmpty(currentSceneId);
+                state.currentScene = null;
+            }
+
+            // UI not set for current scene yet
+            if (state.isPendingSceneUI)
+            {
+                // we get current scene reference
+                state.currentScene ??= GetCurrentScene(currentSceneId, state.loadedScenes);
+
+                // we apply current scene UI
+                if (state.currentScene != null)
+                {
+                    if (ApplySceneUI(state.internalUiContainerComponent, state.uiDocument, state.currentScene))
+                    {
+                        state.isPendingSceneUI = false;
+                    }
+                }
+            }
+        }
+
+        internal static void ApplyParenting(IInternalECSComponent<InternalUiContainer> internalUiContainerComponent)
+        {
+            // check for orphan ui containers
+            var allContainers = internalUiContainerComponent.GetForAll();
+            for (int i = 0; i < allContainers.Count; i++)
+            {
+                var uiContainerData = allContainers[i].value;
+
+                // skip ui containers on root entity since no parenting is needed
+                if (uiContainerData.entity.entityId == SpecialEntityId.SCENE_ROOT_ENTITY)
+                    continue;
+
+                // skip containers with parent already set
+                if (uiContainerData.model.parentElement != null)
+                    continue;
+
+                InternalUiContainer parentDataModel =
+                    internalUiContainerComponent.GetFor(uiContainerData.scene, uiContainerData.model.parentId)?.model;
+
+                // create root entity ui container if needed
+                if (parentDataModel == null && uiContainerData.model.parentId == SpecialEntityId.SCENE_ROOT_ENTITY)
```

Comment:
you can extract to a method and remove comment: `GetOrCreateRootUiContainer`

Reply:
applied suggestion here: https://github.com/decentraland/unity-renderer/pull/3179/commits/223c02f15f101e6c89ae7a6a0dd72d3d72038fa9